linkerd/protocol/http/src/main/scala/com/twitter/finagle/buoyant/linkerd/Headers.scala

```diff
@@ -1,46 +1,232 @@
+      new Stack.Module0[ServiceFactory[Request, Response]] {
+        val role = Stack.Role("DeadlineInject")
+        val description = "Injects deadlines into http headers"
+
+        // TODO use Dtab.ClientFilter once this module can replace finagle's dtab encoding logic.
+        val deadline = new Deadline.ClientFilter
+
+        def make(next: ServiceFactory[Request, Response]) =
+          deadline.andThen(next)
+      }
+
+    val Prefix = Headers.Prefix + "ctx-"
+
+    object Trace {
+      val Key = Prefix + "trace"
+
+      /**
+       * Get a trace id from a base64 encoded buffer.
+       *
+       * Based on com.twitter.finagle.tracing.Trace.idCtx.tryUnmarshal
+       *
+       * The wire format is (big-endian):
+       *   ''reqId:8 parentId:8 traceId:8 flags:8''
+       */
+      def read(b64: String): Try[TraceId] =
+        Try { Base64.getDecoder.decode(b64) }.flatMap(TraceId.deserialize(_))
+
+      def get(headers: HeaderMap): Option[TraceId] =
+        for {
+          header <- headers.get(Key)
+          traceId <- read(header).toOption
+        } yield traceId
+
+      def set(headers: HeaderMap, id: TraceId): Unit = {
+        val bytes = TraceId.serialize(id)
+        val b64 = Base64.getEncoder.encodeToString(bytes)
+        val _ = headers.set(Key, b64)
+      }
+
+      def clear(headers: HeaderMap): Unit = {
+        val _ = headers.remove(Key)
+      }
     }
 
-    def clear(headers: HeaderMap): Unit = {
-      val _ = headers.remove(Key)
+    object Deadline {
+      val Key = Prefix + "deadline"
+
+      def read(v: String): Try[FDeadline] = Try {
+        val values = v.split(' ')
+        val timestamp = Time.fromNanoseconds(values(0).toLong)
+        val deadline = Time.fromNanoseconds(values(1).toLong)
+        FDeadline(timestamp, deadline)
+      }
+
+      def get(headers: HeaderMap): Option[FDeadline] =
+        headers.getAll(Key).foldLeft[Option[FDeadline]](None) { (d0, v) =>
+          (d0, read(v).toOption) match {
+            case (Some(d0), Some(d1)) => Some(FDeadline.combined(d0, d1))
+            case (d0, d1) => d0.orElse(d1)
+          }
+        }
+
+      def write(d: FDeadline): String =
+        s"${d.timestamp.inNanoseconds} ${d.deadline.inNanoseconds}""
+
+      def set(headers: HeaderMap, deadline: FDeadline): Unit = {
+        val _ = headers.set(Key, write(deadline))
+      }
+
+      def clear(headers: HeaderMap): Unit = {
+        val _ = headers.remove(Key)
+      }
+
+      /**
+       * Extract the deadline from the request and, if it exists, use
+       * either that deadline or the one already on the local context,
+       * whichever is stricter.
+       */
+      class ServerFilter extends SimpleFilter[Request, Response] {
+        def apply(req: Request, service: Service[Request, Response]) =
+          get(req.headerMap) match {
+            case None => service(req)
+            case Some(reqDeadline) =>
+              clear(req.headerMap)
```

Comment:
not sure we _need_ to do this, cleared so that it's not passed-through automatically and the client-side is responsible for encoding any downstream deadlines.